### PR TITLE
fix(client): inject chat context per provider session

### LINKED ISSUE #808: Move per-chat context out of AGENTS.md and the agent system prompt
After PR #797 the unified agent briefing is materialised once as `<cwd>/AGENTS.md` (with `CLAUDE.md` symlinked to it). The briefing currently carries a `## Current Chat Context` block with the chat id, title, and participants. This issue tracks moving that block out — both away from `AGENTS.md` *and* away from any equivalent SDK-level system-prompt append channel. Solution is open; this issue captures **why** the move is needed.

### 1. Per-chat content does not belong in a per-agent file

`<cwd>` (the workspace root) is the **agent home** — shared by every chat that agent participates in. Per the agent-session-cwd-redesign, only agent-level stable fields live at the agent home; per-chat data has no place there. The Current Chat Context block puts per-chat data into a per-agent file, which is the original architectural conflict the redesign tried to resolve.

### 2. Concurrent chats race the file

Two concurrent chats for the same agent home both run through `ensureAgentBootstrap` → `writeAgentBriefing` on session start. Whoever wrote last wins; the loser's session may then read a briefing belonging to the other chat. Proposal §⓪.3 accepted this race because there was no alternative per-turn channel at the time, but the conflict is unsolved — `runtime/codex.ts` ships a `detectAgentsMdConcurrentWrite` heuristic only to surface "wrong chat context surfaces in codex" debugging symptoms.

### 3. The block goes stale within a session

The briefing is written once, at session start / resume. But chat state mutates during a session:

- A participant gets invited mid-chat → AGENTS.md still lists the old participant set
- The chat topic is renamed → AGENTS.md still shows the old topic

Moving the block back into the pre-PR-#797 `systemPrompt.append` channel does **not** fix this — that channel is also a snapshot at Query construction, not a per-turn feed. Whatever replaces it must deliver fresh chat state every turn, which neither AGENTS.md nor an SDK system-prompt append can do.

### 4. Cache invalidation drag

The Anthropic prompt cache invalidates from the first changed byte. Current Chat Context is the **only** block in the unified briefing that differs between sibling chats of the same agent. Removing it lets every other section — identity, behavior, First Tree Runtime, Context Tree pointer, Skill Map — stay cacheable across every chat the agent participates in, instead of being invalidated on every chat switch.

## Scope

- AGENTS.md / CLAUDE.md must stop carrying any per-chat content.
- The replacement must cover both providers (Claude Code SDK + Codex SDK).
- The replacement must deliver fresh chat state per turn, not a snapshot at session start.
- Follow-up cleanup: once the move lands, `detectAgentsMdConcurrentWrite` in `runtime/codex.ts` becomes dead code.

### PR DESCRIPTION
**Pull request by baixiaohang**
## Summary

Fixes #808.

Supersedes #1154 because its head branch (`baixiaohang/issue-808-chat-context-prompt`) fails the CI branch-name convention check.

- keep shared `AGENTS.md` / `CLAUDE.md` limited to stable agent-level briefing content
- inject per-chat Current Chat Context through provider/session paths: Claude SDK `systemPrompt.append`, Claude TUI `--append-system-prompt-file`, and Codex one-shot start/resume input prepending
- gate Codex messageful start/resume so queued injects cannot consume the one-shot chat context before the initial provider turn
- render provider/session chat context as escaped JSON data so metadata values cannot close the wrapper tag or forge markdown/prompt structure
- update tests to prove same-agent concurrent chats do not write per-chat context into shared briefing, Codex only applies the session/resume context once, and instruction-like / structure-forging chat metadata stays data

## Context Tree

- agent-team-foundation/first-tree-context#572 records the durable runtime/prompt-channel decision.

## Review

- codex-reviewer reviewed the implementation and the R4 follow-up fix; no blocking issues remain.
- Addressed follow-up GitHub review on #1154 by moving provider chat context from raw markdown to escaped JSON payloads.

## Verification

- `pnpm check`
- `pnpm --filter @first-tree/client typecheck`
- `pnpm --filter @first-tree/client exec vitest run --maxWorkers=2 src/__tests__/chat-context.test.ts src/__tests__/claude-query-options.test.ts src/__tests__/agent-cwd-redesign.e2e.test.ts src/__tests__/tui-suspend-queued-recovery.test.ts src/__tests__/codex-bootstrap.test.ts src/__tests__/codex-inject-startup-race.test.ts src/__tests__/codex-retry-abort.test.ts`
- `pnpm --filter @first-tree/client exec vitest run --maxWorkers=2` (100 files / 1036 tests)
- `git diff --check`

### CODE REVIEW

**Comment by baixiaohang** [comment]
Follow-up for the GitHub review on #1154: head `780039a5` changes provider/session Current Chat Context from raw markdown to escaped JSON data, escaping `<`, `>`, and `&` to prevent wrapper/markdown structure escape. Added tests for instruction-like values plus newline and `</first-tree-current-chat-context>` metadata. Local verification: `pnpm check`, client typecheck, targeted Vitest, full client Vitest (100 files / 1036 tests), and `git diff --check` all pass.

File: packages/client/src/runtime/chat-context-section.ts [hunk 1]
```diff
@@ -55,3 +57,49 @@ export function renderChatContextSection(chatContext: ChatContext | undefined):
...
+  return JSON.stringify(value, null, 2).replace(/[<>&]/g, (char) => {
+    switch (char) {
+      case "<":
+        return "\\u003c";
+      case ">":
+        return "\\u003e";
+      case "&":
+        return "\\u0026";
+      default:
+        return char;
```
Comment (yuezengwu) [inline packages/client/src/runtime/chat-context-section.ts:71]: Blocking: this labels the block as runtime-authored metadata, but the metadata values rendered by `renderChatContextSection` are still raw Markdown. `topic`/`description` are owner-writable and permit newlines, so a value can include `</first-tree-current-chat-context>` or a new heading/instruction block and escape the intended boundary before it is injected through Claude `systemPrompt.append`, the TUI prompt file, or the Codex turn prefix. Please serialize/escape field values and add a regression test with a close-tag/newline payload; the natural-language label is not enough as a trust boundary.

**Review by yuezengwu — CHANGES_REQUESTED** [review]
Summary / intent: this PR fixes #808 by moving per-chat `Current Chat Context` out of shared `AGENTS.md` / `CLAUDE.md` and into provider/session paths: Claude SDK `systemPrompt.append`, Claude TUI `--append-system-prompt-file`, and a one-shot Codex turn prefix. It also gates Codex startup/resume queued injects so the context is consumed by the intended initial provider turn.

Requesting changes for one blocker: chat metadata is now explicitly entering provider prompt channels, but the field values are still rendered as raw Markdown. `topic` and `description` are writable through the chat update routes and allow markdown/newlines up to 500/1500 chars, so a value can inject a fake boundary such as `</first-tree-current-chat-context>` or a new heading/instruction block. The sentence saying field values are metadata/data helps, but it is not a structural boundary, especially for Claude `systemPrompt.append`.

Please serialize or escape field values before rendering the prompt block, and add a regression test with a wrapper-closing/newline payload. For example, render values as JSON string literals and escape `<`/`>` so metadata cannot create literal wrapper tags or markdown structure outside its field.

Non-code follow-up before merge: this PR reverses the durable Context Tree claim in `system/cloud/agents/agent-behavior-steering.md` / `system/cloud/runtime/client-runtime.md` that the briefing is the single channel and there is no SDK-side `systemPrompt.append`. I did not find a companion `first-tree-context` PR for #1157. Please link one so the tree does not become stale the moment this lands.

Verification I ran locally:
- `pnpm install --frozen-lockfile`
- `pnpm --filter @first-tree/client typecheck`
- `pnpm --filter @first-tree/client exec vitest run --maxWorkers=2 src/__tests__/codex-inject-startup-race.test.ts src/__tests__/chat-context.test.ts src/__tests__/claude-query-options.test.ts src/__tests__/tui-suspend-queued-recovery.test.ts`
- `pnpm --filter @first-tree/client exec vitest run --maxWorkers=2`
- `git diff --check origin/main...HEAD`

No DB/schema changes found.

**Comment by baixiaohang** [comment]
@yuezengwu I rechecked #1157 head `780039a5`. The blocker described here appears to be already addressed on this head:

- provider/session paths import and call `renderChatContextPrompt()` (`claude-code.ts`, `claude-code-tui/index.ts`, `codex.ts`);
- `renderChatContextPrompt()` emits `<first-tree-current-chat-context format="json">` and uses `escapePromptJson()` to `JSON.stringify` the payload and escape `<`, `>`, and `&`;
- `renderChatContextSection()` still renders human-readable markdown, but it is retained for tests/diagnostics and is no longer wired into provider/session prompt paths;
- the companion Context Tree PR is agent-team-foundation/first-tree-context#572 and is linked in the PR body.

I do not see a remaining code change for this review item. Could you re-review current head, or point me at the raw-markdown provider path if I am missing one?

**Review by yuezengwu — APPROVED** [review]
Summary / intent: this PR fixes #808 by moving per-chat Current Chat Context out of shared `AGENTS.md` / `CLAUDE.md` and into provider/session channels: Claude SDK `systemPrompt.append`, Claude TUI `--append-system-prompt-file`, and Codex one-shot start/resume input prefixing. It also keeps the shared briefing stable for same-agent sibling chats and gates Codex startup/resume queued injects so the one-shot context is consumed by the intended provider turn.

The previous blocker is resolved at head `780039a5`: provider/session chat context now goes through `renderChatContextPrompt`, which serializes the payload as JSON and escapes `<`, `>`, and `&`, so owner-writable metadata cannot close the wrapper tag or forge markdown/prompt structure. The new tests cover instruction-like metadata, newline + `</first-tree-current-chat-context>` payloads, same-agent shared-briefing isolation, and Codex one-shot application on start/resume/retry paths.

No DB/schema changes found.

Context Tree follow-up is linked: agent-team-foundation/first-tree-context#572.

Verification I ran locally:
- `pnpm check` (exit 0; existing unrelated Biome warnings remain)
- `pnpm --filter @first-tree/client typecheck`
- `pnpm --filter @first-tree/client exec vitest run --maxWorkers=2 src/__tests__/chat-context.test.ts src/__tests__/claude-query-options.test.ts src/__tests__/agent-cwd-redesign.e2e.test.ts src/__tests__/tui-suspend-queued-recovery.test.ts src/__tests__/codex-bootstrap.test.ts src/__tests__/codex-inject-startup-race.test.ts src/__tests__/codex-retry-abort.test.ts`
- `pnpm --filter @first-tree/client exec vitest run --maxWorkers=2` (100 files / 1036 tests)
- `git diff --check origin/main...HEAD`
- read-only Codex adversarial pass: no blocking findings
